Make compiletest do exact matching on triples

varkor opened the pull request:
This avoids the issues of the previous substring matching, ensuring `ARCH_TABLE` and `OS_TABLE` will no longer contain redundant entries. Fixes #48893.

r? @rkruppe

sanxiyn commented:
I don't think this PR is correct, because we already rely on substring matching. For example, with this change, `arm-linux-androideabi` will not match `ignore-android`.

varkor commented:
@sanxiyn: Ah, that's awkward — do you know how prevalent this is (that is: using substring matching)? Would it be plausible to special case strings like `android` instead?

sanxiyn commented:
I think you should review the output of `rustc --print target-list` and add missing entries to `ARCH_TABLE` and `OS_TABLE`. It should be doable although tedious.

rkruppe commented:
Putting all the minor variations of the triple into compiletest probably also makes adding new targets (incl. minor variations of existing targets) more annoying, but hopefully less so than learning months later your `ignore-*` lines don't do what they appear to be doing.

varkor commented:
To clarify: you think the tests should be all updated too, so that:
`// ignore-android` becomes:
```
// ignore-android
// ignore-androideabi
```
I.e. no more substring matching whatsoever.

rkruppe commented:
No, that's not what I'm saying. Triples contain more information (and in some cases, random noise) than we need for the purpose of "ignore this test on this architecture". For example, consider how `i386`, `i586` and `i686` triples are all considered for `ignore-x86` currently, or how `arm64` in triples is treated as an alias for `aarch64`. I'm in favor of extending that canonicalization to other variations in arch and OS components. This still requires someone to go through all the triples, but only in compiletest, not in every test ever (where it would add no useful information).

This is a judgement call to some degree, there may be distinctions in triples that we may want to preserve. But we'll learn about those when someone wants to make that distinction in a test.

varkor commented:
Okay, I've added the OSes and architectures that weren't previously considered, and added what I think are reasonable canonicalisations for each of them — they should match the previous substring matching unless something very specific was being done before (which is probably most easily observed by letting bors test it).

rkruppe commented on src/tools/compiletest/src/util.rs:
```diff
@@ -14,43 +14,62 @@ use common::Config;
     ("dragonfly", "dragonfly"),
+    ("eabi", "eabi"),
+    ("eabihf", "eabi"),
+    ("elf", "elf"),
```
Uh these seem rather odd. I know triples can be weird sometimes, are these actual "operating systems" or do they just happen to be in that position in the triple?

rkruppe replied:
Um, I guess a better question is: which triples do there come from?

varkor replied:
They did seem odd, but I didn't feel certain enough to exclude them. They come from`thumbv6m-none-eabi` (and the other `thumb*` triples) and `msp430-none-elf`. Do you think I should just exclude these?

rkruppe replied:
Ah so these triples have no "vendor" component, so the *second* component is the "OS" and the last component is the *ABI*.

varkor replied:
Ah — is it always safe to make this assumption? (I.e. elided component is vendor?)

bors commented:
@varkor: :key: Insufficient privileges: Not in reviewers

bors commented:
@varkor: :key: Insufficient privileges: not in try users

varkor commented:
Not sure why bors thought I was trying anything funny... I know my place.

bors commented:
@varkor: :key: Insufficient privileges: Not in reviewers

bors commented:
@varkor: :key: Insufficient privileges: not in try users

rkruppe commented on src/tools/compiletest/src/header.rs:
```diff
@@ -574,12 +574,14 @@ impl Config {
-                name == util::get_pointer_width(&self.target) ||    // pointer width
-                name == self.stage_id.split('-').next().unwrap() || // stage
-                Some(name) == util::get_env(&self.target) ||        // env
+                util::matches_os(&self.target, triple.get(2), name) || // target
```
I was always worried about "hard-coding" the meaning of triple components like this since they are somewhat flexible. Since it turns out we even have triples in-tree for which this breaks down, let's not do that...

bors commented:
@varkor: :key: Insufficient privileges: Not in reviewers

bors commented:
@varkor: :key: Insufficient privileges: not in try users

rkruppe commented on src/tools/compiletest/src/util.rs:
```diff
@@ -14,70 +14,90 @@ use common::Config;
+            return name == "emscripten" || name == "wasm32-bare"
+        }
+        for &(triple_os, os) in OS_TABLE {
+            if target_os == &triple_os {
```
... instead let's simply look at all triple components (like it previously did, just without the substring search now). Namespace collisions between the different triple components are practically impossible and would cause far worse problems in the ecosystem of "things that consume triples".

varkor replied:
Sure, I wasn't sure if the components could ever coïncide — but I shall do that instead.

pietroalbini commented:
@varkor it was actually a bors bug, don't worry :)

sanxiyn commented:
@bors r+

bors commented:
:pushpin: Commit dfaecbb has been approved by `sanxiyn`

<!-- @bors r=sanxiyn dfaecbbe5bb3b57a0a55833f8b8a57291dc5eded -->

rkruppe commented on src/tools/compiletest/src/header.rs:
```diff
@@ -574,12 +574,14 @@ impl Config {
-                name == util::get_pointer_width(&self.target) ||    // pointer width
-                name == self.stage_id.split('-').next().unwrap() || // stage
-                Some(name) == util::get_env(&self.target) ||        // env
+                util::matches_os(&self.target, &triple, name) || // target
```
Nit: I think now that these functions look at the whole triple again, there's not much to be gained from splitting here and passing the component explicitly (especially for `matches_os` which needs the whole triple anyway). Perhaps just revert these changes and split in the `util::*` functions to keep the interface simpler?

rkruppe commented on src/tools/compiletest/src/header.rs:
```diff
@@ -574,12 +574,14 @@ impl Config {
+                name == util::get_arch(&triple) ||               // architecture
+                name == util::get_pointer_width(&self.target) ||       // pointer width
+                name == self.stage_id.split('-').next().unwrap() ||    // stage
+                Some(&name) == triple.get(3) ||                        // env
```
Nit: This is a pure refactoring, right? If so, I'd prefer to leave it out of this PR, for consistency. (Splitting multiple times is sure to have completely neglegible performance impact.)

varkor commented:
@rkruppe: bors can wait! I've squashed most of the commits down and simplified the changes.

rkruppe commented:
Thanks! 

@bors r+

bors commented:
:pushpin: Commit 61e1fbc has been approved by `rkruppe`

<!-- @bors r=rkruppe 61e1fbc659103513e68eae29ea830d798e2ec2d5 -->

bors commented:
:umbrella: The latest upstream changes (presumably #49264) made this pull request unmergeable. Please resolve the merge conflicts.
